### PR TITLE
Docker - Support Docker pull

### PR DESCRIPTION
**Pull request by Or-Gabay**
- [x] All [tests](https://ci.appveyor.com/project/jfrog-ecosystem/jenkins-artifactory-plugin) passed. If this feature is not already covered by the tests, I added new tests.
- [x] This pull request is created in the [jfrog/jenkins-artifactory-plugin](https://github.com/jfrog/jenkins-artifactory-plugin) repository.
-----
 Support for scripted & declarative docker pull, including:
* Fat-manifest.
* Marker layers.
* Collect build info.

Related PR:
 * https://github.com/jfrog/build-info/pull/397

### CODE REVIEW

File: src/test/java/org/jfrog/hudson/pipeline/integration/CommonITestsPipeline.java [hunk 1]
```diff
@@ -511,6 +510,30 @@ void dockerPushTest(String buildName) throws Exception {
...
+        // Get image name
+        String domainName = System.getenv("JENKINS_ARTIFACTORY_DOCKER_PULL_DOMAIN");
+        if (StringUtils.isBlank(domainName)) {
+            throw new MissingArgumentException("The JENKINS_ARTIFACTORY_DOCKER_PULL_DOMAIN environment variable is not set.");
+        }
+        if (!StringUtils.endsWith(domainName, "/")) {
+            domainName += "/";
+        }
+        String imageName = domainName + "hello-world:latest";
+        String host = System.getenv("JENKINS_ARTIFACTORY_DOCKER_HOST");
```
Comment (yahavi) [inline src/test/java/org/jfrog/hudson/pipeline/integration/CommonITestsPipeline.java:524]:
```suggestion
```

File: src/test/java/org/jfrog/hudson/pipeline/integration/CommonITestsPipeline.java [hunk 2]
```diff
@@ -511,6 +510,30 @@ void dockerPushTest(String buildName) throws Exception {
         }
     }
 
+    void dockerPullTest(String buildName) throws Exception {
+        Assume.assumeFalse("Skipping Docker tests", SystemUtils.IS_OS_WINDOWS);
+        // Get image name
```
Comment (yahavi) [inline src/test/java/org/jfrog/hudson/pipeline/integration/CommonITestsPipeline.java:515]:
```suggestion
        // Assert 'JENKINS_ARTIFACTORY_DOCKER_PULL_DOMAIN' environment variable exist
```

File: src/test/java/org/jfrog/hudson/pipeline/integration/CommonITestsPipeline.java [hunk 3]
```diff
@@ -511,6 +510,30 @@ void dockerPushTest(String buildName) throws Exception {
...
+        String host = System.getenv("JENKINS_ARTIFACTORY_DOCKER_HOST");
+        // Run pipeline
+        runPipeline("dockerPull", false);
+        String buildNumber = "1";
+        // Get build info
+        Build buildInfo = getBuildInfo(buildInfoClient, buildName, buildNumber);
+        assertEquals(1, buildInfo.getModules().size());
+        List<Module> modules = buildInfo.getModules();
+        Module module = modules.get(0);
+        assertEquals(null, module.getArtifacts());
```
Comment (yahavi) [inline src/test/java/org/jfrog/hudson/pipeline/integration/CommonITestsPipeline.java:533]:
```suggestion
        assertNull(module.getArtifacts());
```

File: src/test/java/org/jfrog/hudson/pipeline/integration/CommonITestsPipeline.java [hunk 4]
```diff
@@ -511,6 +510,30 @@ void dockerPushTest(String buildName) throws Exception {
...
+        assertTrue(module.getDependencies().size() > 0);
```
Comment (yahavi) [inline src/test/java/org/jfrog/hudson/pipeline/integration/CommonITestsPipeline.java:534]: I suggest to check that the actual image exist with the image name, if possible

File: src/main/java/org/jfrog/hudson/pipeline/scripted/steps/DockerPullStep.java [hunk 5]
```diff
@@ -74,15 +79,18 @@ protected BuildInfo run() throws Exception {
                 getContext().onFailure(new MissingArgumentException("Missing 'image' parameter"));
```
Comment (yahavi) [inline src/main/java/org/jfrog/hudson/pipeline/scripted/steps/DockerPullStep.java:79]: Please avoid MissingArgumentException. See: https://github.com/jfrog/jenkins-artifactory-plugin/pull/327
```suggestion
                getContext().onFailure(new AbortException("Missing 'image' parameter"));
```

File: src/main/java/org/jfrog/hudson/pipeline/scripted/steps/DockerPullStep.java [hunk 6]
```diff
@@ -74,15 +79,18 @@ protected BuildInfo run() throws Exception {
...
                 return null;
             }
-
+            if (step.getTargetRepo() == null) {
+                getContext().onFailure(new MissingArgumentException("Missing 'targetRepo' parameter"));
```
Comment (yahavi) [inline src/main/java/org/jfrog/hudson/pipeline/scripted/steps/DockerPullStep.java:83]:
```suggestion
                getContext().onFailure(new AbortException("Missing 'targetRepo' parameter"));
```

**Review by yahavi — CHANGES_REQUESTED** [review]

File: src/main/java/org/jfrog/hudson/pipeline/common/types/Docker.java [hunk 7]
```diff
@@ -82,14 +81,15 @@ public void push(Map<String, Object> dockerArguments) {
     }
 
     @Whitelisted
-    public void pull(String imageTag) {
```
[the remark this answers is not in the document]
Comment (Or-Gabay) [reply src/main/java/org/jfrog/hudson/pipeline/common/types/Docker.java:85]: We are not able to collect build info if there is no repository. There is no documentation for that command so have no good reason to keep supporting that.

File: src/main/java/org/jfrog/hudson/pipeline/common/types/Docker.java [hunk 8]
```diff
@@ -82,14 +81,15 @@ public void push(Map<String, Object> dockerArguments) {
...
-        pull(imageTag, null);
+    public void pull(String imageTag, String targetRepository) {
+        pull(imageTag, targetRepository, null);
     }
 
     @Whitelisted
-    public void pull(String imageTag, BuildInfo providedBuildInfo) {
```
[the remark this answers is not in the document]
Comment (Or-Gabay) [reply src/main/java/org/jfrog/hudson/pipeline/common/types/Docker.java:90]: https://github.com/jfrog/jenkins-artifactory-plugin/pull/347#discussion_r509948493

File: src/main/java/org/jfrog/hudson/pipeline/declarative/steps/docker/DockerPullStep.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    private String serverId;
+    private String image;
+    private String host;
+    private String buildNumber;
+    private String buildName;
+    private String targetRepo;
+    private String javaArgs;
+
+    @DataBoundConstructor
+    public DockerPullStep(String serverId, String image, String targetRepo, String javaArgs) {
```
[the remark this answers is not in the document]
Comment (Or-Gabay) [reply src/main/java/org/jfrog/hudson/pipeline/declarative/steps/docker/DockerPullStep.java:29]: Removed

File: src/main/java/org/jfrog/hudson/pipeline/declarative/steps/docker/DockerPullStep.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    @DataBoundSetter
+    public void setJavaArgs(String javaArgs) {
+        this.javaArgs = javaArgs;
+    }
+
+    public static class Execution extends ArtifactorySynchronousNonBlockingStepExecution<Void> {
+
+        private transient DockerPullStep step;
```
[the remark this answers is not in the document]
Comment (Or-Gabay) [reply src/main/java/org/jfrog/hudson/pipeline/declarative/steps/docker/DockerPullStep.java:58]: For some reason, the credentials inside ```step``` won't pass to the agent node.

**Pull request merged**